### PR TITLE
replenish prompts directly after accumulating

### PR DESCRIPTION
**Pull request by mnoukhov**
successful [single GPU on beaker](https://beaker.allen.ai/orgs/ai2/workspaces/open-instruct-dev/work/01KA21TQT0YGQFJ5EPM8CF0JX9?taskId=01KA21TQT642Q8D9AWQZBPZCJW&jobId=01KA21TQY6SYPK15FAJ0QXDJ1C) 

moves prompt loading into `accumulate_inference_batches`:
- when `replenish_prompts=True` we add a prompt on the prompt_Q after consuming a finished result

simplifies a lot of our batch and dataset stuff
- now pulls single examples instead of `batch_size` 
- adds each prompt, one by one to the generator's queue
- removes extra code and functions

---

> [!NOTE]
> Moves to single-example prompt flow with immediate replenishment during accumulation, removes batch-splitting helpers, and updates training/eval and tests accordingly.
> 
> - **Core sampling/accumulation**:
>   - Add `replenish_prompts` option to `accumulate_inference_batches`; when enabled, pulls next index from `ShufflingIterator` and enqueues a new prompt via `add_prompt_to_generator` after each completed result.
>   - Introduce `add_prompt_to_generator(example, idx, ...)` to insert prompts and track them in `PendingQueriesMap`.
>   - Remove `next_batch` and `split_and_insert_batch`; flow now operates on single examples.
>   - `ShufflingIterator.__next__` returns a single index when `batch_size==1`.
> - **Training loop**:
>   - Seed initial work by enqueuing `async_steps * num_unique_prompts_rollout` single prompts.
>   - Data prep thread now passes `replenish_prompts=True` and receives/train-packs batches continuously.
>   - Replace `eval_batch` with iterating over `eval_dataset`; enqueue each eval example individually.
>   - Build iterator with `batch_size=1`; simplify refill logic and remove manual prompt refill counters.
> - **API/args tweaks**:
>   - Extend `accumulate_inference_batches` signature (dataset, queues, iterator, training_step); assert presence when `replenish_prompts` or `no_resampling_pass_rate` used.
>   - Minor variable renames (`dataset` -> `dataset_name`), add `Dataset` import.
> - **Tests**:
>   - Replace batch-splitting usages with `setup_and_add_prompts_to_generator` and direct `add_prompt_to_generator` calls.
>   - Remove `model_utils` dependency in tests; update expectations for per-prompt requests and multiple-sample reference counting.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 163d8207aec6411bda27404daede01aa68e74e1e. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: open_instruct/grpo_fast.py [hunk 1]
```diff
@@ -673,7 +659,8 @@ def __init__(self, data: np.ndarray, batch_size: int, seed: int | None = None):
     def __iter__(self) -> Iterator[list[int]]:
         return self
 
-    def __next__(self) -> list[int]:
+    def __next__(self) -> list[int] | int:
+        """Return a list of next indices or a single index if batch size is 1"""
```
Comment (finbarrtimbers) [inline open_instruct/grpo_fast.py:664]: will batch size ever be greater than 1?

File: open_instruct/grpo_fast.py [hunk 2]
```diff
@@ -1786,6 +1802,7 @@ def accumulate_inference_batches(
             logging.debug(
                 f"[Data Preparation Thread] Filtered prompt with reward std 0, total filtered {total_filtered_prompts}"
             )
+
```
Comment (finbarrtimbers) [inline open_instruct/grpo_fast.py]: remove plz

File: open_instruct/grpo_fast.py [hunk 3]
```diff
@@ -1733,6 +1735,20 @@ def accumulate_inference_batches(
 
         query, ground_truth, dataset, raw_query = pending_queries_map.pop(result.dataset_index)
 
+        # Replenish generation queue with new prompt
+        if replenish_prompts:
+            dataset_index = next(iter_dataloader)
+            add_prompt_to_generator(
+                dataset[dataset_index],
```
[the remark this answers is not in the document]
Comment (mnoukhov) [reply open_instruct/grpo_fast.py]: fixed

File: open_instruct/grpo_fast.py [hunk 4]
```diff
@@ -1786,6 +1804,7 @@ def accumulate_inference_batches(
             logging.debug(
                 f"[Data Preparation Thread] Filtered prompt with reward std 0, total filtered {total_filtered_prompts}"
             )
+
```
[the remark this answers is not in the document]
Comment (mnoukhov) [reply open_instruct/grpo_fast.py]: Uh no, prompts are consumed regardless of filtering.

For each prompt we pull from finished generations, we put one to be generated, simple as that

File: open_instruct/grpo_fast.py (hunk 1 above)
Comment (mnoukhov) [reply open_instruct/grpo_fast.py:664]: it could be, theoretically, but with this code it never is

should I remove that functionality?

File: open_instruct/grpo_fast.py (hunk 2 above)
Comment (mnoukhov) [reply open_instruct/grpo_fast.py]: done
